MapboxCoreNavigation/RouteController.swift

```diff
@@ -676,6 +679,44 @@ extension RouteController: CLLocationManagerDelegate {
         }
     }
     
+    func userExitedRouteAtIntersection(_ location: CLLocation) -> Bool {
+        guard var intersections = routeProgress.currentLegProgress.currentStepProgress.step.intersections else { return false }
+        let currentStepProgress = routeProgress.currentLegProgress.currentStepProgress
+        
+        // The intersections array does not include the upcoming maneuver intersection.
+        if let upcomingStep = routeProgress.currentLegProgress.upComingStep, let upcomingIntersection = upcomingStep.intersections, let firstUpcomginIntersection = upcomingIntersection.first {
+            intersections = intersections + [firstUpcomginIntersection]
+        }
+        
+        routeProgress.currentLegProgress.currentStepProgress.intersectionsIncludingUpcomingManeuverIntersection = intersections
+        
+        // Don't search for previous intersections
+        for intersection in intersections.suffix(from: currentStepProgress.intersectionIndex) {
+            let dist = distance(along: currentStepProgress.step.coordinates!, from: location.coordinate, to: intersection.location)
+            // Check if the user is at an intersection
+            if dist <= RouteControllerManeuverZoneRadius / 2 {
+                guard let newCurrentIntersection = routeProgress.currentLegProgress.currentStepProgress.currentIntersection else { return false }
+                
+                // Loop over all possible headings at an intersection.
```

Comment:
Rather than looking at specific outlet headings, it would be simpler to decrease the rerouting radius thresholds when the user is near an intersection. Would that yield acceptable results?

I’m just concerned that we’re relying too heavily on outlet headings, but road headings are less reliable near large intersections in general, due to how they’re modeled in OpenStreetMap. But as long as we’re handling intersections like the following examples well, then this approach could be fine:

1. [Road divides at intersection](https://www.mapbox.com/get-directions/#19/37.35013/-121.95944?coordinates=-121.959435,37.349865;-121.95936,37.350293)
2. [Road curves sharply at intersection](https://www.mapbox.com/get-directions/#18.51/39.3361/-84.27812?coordinates=-84.27780130105117,39.33588655554837;-84.27811584679313,39.33609560406964)
3. [Road curves gently at intersection](https://www.mapbox.com/get-directions/#17.18/39.10636/-84.5548?coordinates=-84.554801,39.106355;-84.554361,39.106354)

In the third example, the heading for Wilder Avenue is reported as 90°. But if the user happens to travel at such a speed that a location update comes in while the user’s course is still 135° or already 45°, then we’ll incorrectly reroute, right? Or is something else preventing that from happening?